File: apps/web/petProject/src/types/shop.ts

interface ShopItem {
    shopID: string;
    itemID: string;
    price: number;
    quantity: number | null;
};

function toDbShopItem(si: ShopItem): dbShopItem{
    return {
        item_id: si.itemID,
        price: si.price,
        shop_id: si.shopID,
        quantity: si.quantity
    };
}

interface dbShopItem {
    item_id: string;
    price: number;
    quantity: number | null;
    shop_id: string;
};

function toShopItem (si: dbShopItem): ShopItem{
    return {
        itemID: si.item_id,
        shopID: si.shop_id,
        quantity: si.quantity,
        price: si.price
    };
}

function createShopItem(overrides: Partial<ShopItem> = {}): ShopItem {
    return {
        shopID: "",
        itemID: "",
        price: -1,
        quantity: null,
        ...overrides
    };
}


interface Shop {
    id: string;
    name: string;
    items: ShopItem[];
    lastUpdate: Date
};


function toDbShop(shop: Shop): dbShop {
    return {
        shop_id: shop.id,
        name: shop.name,
        last_update: new Date()
    }
}


interface dbShop {
    last_update: Date;
    name: string;
    shop_id: string;
};


function toShop(shop: dbShop): Shop {
    return {
        id: shop.shop_id,
        name: shop.name,
        items: [],
        lastUpdate: shop.last_update
    }
}


function createShop(overrides: Partial<Shop> = {}): Shop {
    return {
        id: "",
        name: "Test Shop",
        items: [],
        lastUpdate: new Date(),
        ...overrides   
    };
}


export type { Shop, ShopItem, dbShop, dbShopItem };
export { createShop, createShopItem, toDbShop, toShop, toDbShopItem, toShopItem };

File: apps/petProject/shared/Models/Shop.swift
//
//  Shop.swift
//  petProject
//
//  Created by Aaron Foster on 3/10/26.
//

import Foundation
import Combine

struct ShopItem: Identifiable, Codable{
    var id: String
    var price: Int
    var quantity: Int?
    var shopID: String
    
    enum CodingKeys: String, CodingKey {
        case id = "item_id"
        case price
        case quantity
        case shopID = "shop_id"
    }
    
    init(id: String = "", price: Int = -1, quantity: Int? = nil, shopID: String = "") {
        self.id = id
        self.price = price
        self.quantity = quantity
        self.shopID = shopID
    }
}

class Shop: Identifiable, Codable {
    var id: String
    var name: String
    var items: [String: ShopItem] = [:]
    var lastUpdate: Date
    
    enum CodingKeys: String, CodingKey {
        case id = "shop_id"
        case name
        case lastUpdate = "last_update"
    }
    
    init(id: String = "test_id", name: String = "Test Shop", inventory: [String: ShopItem] = [:], lastUpdate: Date = Date()){
        self.id = id
        self.name = name
        self.items = inventory
        self.lastUpdate = lastUpdate
    }
    
    func removeItem(_ item: String) {
        let rmvItem = items[item]
        
        guard rmvItem != nil, rmvItem!.quantity != nil else {return}
        if(rmvItem!.quantity! > 0){
            items[item] = ShopItem(id: item, price: rmvItem!.price, quantity: rmvItem!.quantity! - 1, shopID: rmvItem!.shopID)
        }
    }
    
    func allShopItems() -> [ShopItem] {
        return Array(items.values)
    }
}
